Added SSL for console and apiserver

dymurray opened the pull request.

jwmatthews commented:
I didn't see an ability to store the generated certs and reuse for new deployments, i.e. new ec2 instances.


I suggest we see if we can do the following:
1) Check if local cert data exists for the specific hostname, if it does copy it to the new instance
2) As playbook runs, check if cert data exists on the instance, if not then go through the generation
3) Ensure cert data from instance is copied to local host
4) Add .gitignore entries to ignore the cert data on local host  

Ideal flow would be we generate the cert data once and reuse, copying to the new instances so we can destroy/recreate.

Renewing certs would be a little trickier to handle, for purposes of this demo setup, I'd ignore renewal.  We can just delete the cert data and re-run in future.

Consider creating a new role for generating the certs and handling letsencrypt specifics.

dymurray commented:
@jwmatthews I have made some changes around the test-cert flag to now be a variable as well so ensure you don't miss that when testing. Please let me know if all of your points were addressed from your previous comment.

jwmatthews commented:
I saw success with initial test.
site is up and url shows up 'secured' from browser.

Looking good so far.  Will run through some more tests.

jwmatthews commented:
Had success with a full run of demo with this.

@dymurray wondering if we should keep a non-letsencrypt mode of running as well, where we can set a variable to use letsencrypt or else default to the prior behavior.

Thinking about when we next move to running this environment locally on a laptop....can we still do the letsencrypt certs in that case?

jwmatthews commented:
@dymurray let's fix this, re-run of setup_environment

TASK [service_catalog_setup : Create apiserver secret through OC Template] *****
fatal: [34.205.102.146]: FAILED! => {"changed": true, "cmd": "/usr/local/bin/oc create -f /tmp/apiserver-ssl-secret.yaml", "delta": "0:00:00.159857", "end": "2017-04-20 18:31:38.226685", "failed": true, "rc": 1, "start": "2017-04-20 18:31:38.066828", "stderr": "Error from server (AlreadyExists): error when creating \"/tmp/apiserver-ssl-secret.yaml\": secrets \"apiserver-ssl\" already exists", "stdout": "", "stdout_lines": [], "warnings": []}

jmontleon commented on ansible/roles/ssl_setup/tasks/main.yml:
```diff
@@ -0,0 +1,116 @@
+    stat:
+      path: certs/{{ openshift_hostname }}-console-cert.pem
+    register: certsLocal
+    delegate_to: localhost
```
If you add a line to this,  "sudo: no", then it will not error if you're not using passwordless sudo and still appears to work fine.

dymurray replied:
Thank you for this. That's a good tip.

jwmatthews commented:
@dymurray let's wait till Monday before we consider merging this to master.

@jmrodri has asked us to keep master stable throughout today to help him reduce issues with the async testing/merge.

dymurray commented:
@jmontleon @jwmatthews I updated the PR so that you can change 'use_ssl; from True to False and keep running the 'reset_environment.sh' script to switch between ssl and non ssl. The only limitation that exists is that if you have an SSL environment and don't manually 'oc cluster down' the cluster then resetting the environment with non-ssl will not break anything, but the console will still be secure.

This is due to me being unable to use the config file as a template since it's dynamic and created by Openshift. But other than that it doesn't break anything so i don't see it as a problem. I also fixed the problem with the secret being recreated if it already exists.

I agree that we should wait to merge until @jmrodri feels master is stable.

dymurray commented:
I am also making a change to the reset script to rerun 'oc cluster up' to overwrite all of the configs to truly reset the cluster. This allows me to fully switch from SSL to non SSL.

jmontleon commented:
There are a ton of warnings about using 

`
when: "{{` use_ssl }} == True"
`

in these files.

`
../ansible/roles/openshift_setup/tasks/main.yml
../ansible/roles/service_catalog_setup/tasks/main.yml
../ansible/setup_environment.yml
../ansible/reset_environment.yml
`
A quick search and replace fixes these:
`
:%s/\"{{ use_ssl }} == True\"/use_ssl == True/g
`

You also have use_ssl on 70 and 81 wrapped in brackets that should be removed in 

`
../ansible/roles/service_catalog_setup/tasks/main.yml
`

jmontleon commented:
It looks like the advice I gave you to add sudo: no should have been become: no, otherwise running the playbook throws a huge purple deprecation warning when it starts running.

jmontleon commented on ansible/roles/openshift_setup/tasks/main.yml:
```diff
@@ -80,14 +89,122 @@
   - name: Run oc cluster up
     shell: "{{ oc_cmd }} cluster up --routing-suffix={{ openshift_hostname }} --public-hostname={{ openshift_hostname }} --host-pv-dir={{ persistedvol_mount_point }} --version=latest --image=docker.io/ansibleplaybookbundle/origin"
-    when: oc_cluster_up_first_run and not master_config_stat.stat.exists
+    when: (oc_cluster_up_first_run and not master_config_stat.stat.exists) or ssl_changed
```
ssl_changed is never initialized False so it's either undefined or True. On reruns it fails.

You can do and/or `ssl_changed is defined`

dymurray replied:
Thank you good catch.

jmontleon commented on ansible/roles/service_catalog_setup/tasks/main.yml:
```diff
@@ -26,6 +26,10 @@
     register: project
 
+  - name: check if the service-catalog ssl secret exists
+    shell: "{{ oc_cmd }} get secret --no-headers=true | awk '{ print $1 }' | grep -E '^apiserver-ssl( |$)' | cat"
```
needs a `-n service-catalog` to  check the correct namespace otherwise reruns fails.

jmontleon commented on ansible/roles/ssl_setup/tasks/main.yml:
```diff
@@ -0,0 +1,117 @@
+      path: certs/{{ openshift_hostname }}-console-cert.pem
+    register: certsLocal
+    delegate_to: localhost
+    sudo: no
```
`become: no`
Sorry about that.

dymurray commented:
@jmontleon Thank you, I reread more about the use of brackets and totally see where I was confused on that. Good feedback and I fixed your comments.
